File: backend/gemini_client.py
"""Gemini free-tier fallback client for BB and document scanning.

Primary inference remains local Ollama. This module is only used when
OLLAMA_URL is unreachable and GEMINI_API_KEY is set (free tier is sufficient
for HAVEN's civic-support workloads).
"""
from __future__ import annotations

import json
import logging
import os
from typing import Any, Optional

import httpx

logger = logging.getLogger("haven.gemini")

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY", "")
GEMINI_MODEL = os.environ.get("GEMINI_MODEL", "gemini-1.5-flash-latest")
GEMINI_TIMEOUT = float(os.environ.get("GEMINI_TIMEOUT", "60"))


def _gemini_url(action: str = "generateContent") -> str:
    return (
        f"https://generativelanguage.googleapis.com/v1beta/models/{GEMINI_MODEL}"
        f":{action}?key={GEMINI_API_KEY}"
    )


def _build_request_body(
    system_prompt: str,
    user_prompt: str,
    json_mode: bool = False,
    image_base64: Optional[str] = None,
) -> dict[str, Any]:
    parts = [{"text": user_prompt}]
    if image_base64:
        parts.insert(
            0,
            {
                "inline_data": {
                    "mime_type": "image/png",
                    "data": image_base64,
                }
            },
        )

    body: dict[str, Any] = {
        "contents": [
            {"role": "user", "parts": [{"text": system_prompt}]},
            {"role": "user", "parts": parts},
        ],
        "generationConfig": {
            "temperature": 0.7,
            "maxOutputTokens": 1024,
        },
    }
    if json_mode:
        body["generationConfig"]["responseMimeType"] = "application/json"
    return body


async def gemini_chat(
    system_prompt: str,
    user_prompt: str,
    json_mode: bool = False,
    image_base64: Optional[str] = None,
) -> str:
    """Send a request to Gemini and return the text response."""
    if not GEMINI_API_KEY:
        raise RuntimeError("GEMINI_API_KEY not configured")

    body = _build_request_body(system_prompt, user_prompt, json_mode, image_base64)

    async with httpx.AsyncClient(timeout=GEMINI_TIMEOUT) as client:
        r = await client.post(_gemini_url(), json=body)
        r.raise_for_status()
        data = r.json()

    candidates = data.get("candidates", [])
    if not candidates:
        raise RuntimeError("Gemini returned no candidates")

    parts = candidates[0].get("content", {}).get("parts", [])
    if not parts:
        raise RuntimeError("Gemini returned empty content parts")

    return parts[0].get("text", "").strip()


async def gemini_chat_history(messages: list[dict[str, str]]) -> str:
    """Send a multi-turn chat history to Gemini (last message is the prompt)."""
    if not GEMINI_API_KEY:
        raise RuntimeError("GEMINI_API_KEY not configured")

    contents = []
    for m in messages:
        role = m.get("role", "user")
        if role == "system":
            role = "user"
        contents.append({"role": role, "parts": [{"text": m.get("content", "")}]})

    body = {
        "contents": contents,
        "generationConfig": {"temperature": 0.7, "maxOutputTokens": 1024},
    }

    async with httpx.AsyncClient(timeout=GEMINI_TIMEOUT) as client:
        r = await client.post(_gemini_url(), json=body)
        r.raise_for_status()
        data = r.json()

    candidates = data.get("candidates", [])
    if not candidates:
        raise RuntimeError("Gemini returned no candidates")

    parts = candidates[0].get("content", {}).get("parts", [])
    if not parts:
        raise RuntimeError("Gemini returned empty content parts")

    return parts[0].get("text", "").strip()

File: backend/src/lib/geminiClient.ts
import axios from 'axios';
import dotenv from 'dotenv';
import { reliabilityEngine } from '../services/ReliabilityEngine';

dotenv.config();

export interface GeminiMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

export interface GeminiOptions {
  temperature?: number;
  maxOutputTokens?: number;
  topP?: number;
}

const GEMINI_API_KEY = process.env.GEMINI_API_KEY || '';
const GEMINI_MODEL = process.env.GEMINI_MODEL || 'gemini-1.5-flash';
const GEMINI_URL = process.env.GEMINI_URL || 'https://generativelanguage.googleapis.com/v1beta';
const GEMINI_TIMEOUT = parseInt(process.env.GEMINI_TIMEOUT || '45000', 10);

function toGeminiContents(messages: GeminiMessage[]) {
  return messages
    .filter((m) => m.role !== 'system')
    .map((m) => ({
      role: m.role === 'assistant' ? 'model' : 'user',
      parts: [{ text: m.content }],
    }));
}

function extractSystemPrompt(messages: GeminiMessage[]): string {
  return messages
    .filter((m) => m.role === 'system')
    .map((m) => m.content)
    .join('\n\n');
}

export async function geminiChat(
  messages: GeminiMessage[],
  options: GeminiOptions = {}
): Promise<string> {
  if (!GEMINI_API_KEY) {
    throw new Error('Gemini API key is not configured (GEMINI_API_KEY)');
  }

  const url = `${GEMINI_URL}/models/${GEMINI_MODEL}:generateContent?key=${encodeURIComponent(GEMINI_API_KEY)}`;
  const systemPrompt = extractSystemPrompt(messages);
  const contents = toGeminiContents(messages);

  if (contents.length === 0) {
    throw new Error('Gemini requires at least one user/assistant message');
  }

  const response = await reliabilityEngine.execute<any>(
    'gemini.chat',
    () =>
      axios.post(
        url,
        {
          systemInstruction: systemPrompt
            ? {
                parts: [{ text: systemPrompt }],
              }
            : undefined,
          contents,
          generationConfig: {
            temperature: options.temperature ?? 0.7,
            topP: options.topP ?? 0.9,
            maxOutputTokens: options.maxOutputTokens ?? 700,
          },
        },
        {
          timeout: GEMINI_TIMEOUT,
          headers: { 'Content-Type': 'application/json' },
        }
      ),
    {
      timeoutMs: GEMINI_TIMEOUT,
      maxRetries: 2,
      baseDelayMs: 300,
      maxDelayMs: 2500,
      maxConcurrent: 25,
      circuitFailureThreshold: 4,
      circuitHalfOpenAfterMs: 15000,
      circuitSuccessThreshold: 2,
    }
  );

  const text = response?.data?.candidates?.[0]?.content?.parts
    ?.map((p: any) => p?.text)
    ?.filter(Boolean)
    ?.join('\n')
    ?.trim();

  if (!text) {
    throw new Error('Gemini returned empty or invalid response');
  }

  return text;
}

export async function geminiHealth(): Promise<{ ok: boolean; model: string; detail?: string }> {
  if (!GEMINI_API_KEY) {
    return {
      ok: false,
      model: GEMINI_MODEL,
      detail: 'GEMINI_API_KEY is missing',
    };
  }

  try {
    const url = `${GEMINI_URL}/models/${GEMINI_MODEL}:generateContent?key=${encodeURIComponent(GEMINI_API_KEY)}`;
    await reliabilityEngine.execute<any>(
      'gemini.health',
      () =>
        axios.post(
          url,
          {
            contents: [{ role: 'user', parts: [{ text: 'ping' }] }],
            generationConfig: { maxOutputTokens: 8 },
          },
          { timeout: 8000, headers: { 'Content-Type': 'application/json' } }
        ),
      {
        timeoutMs: 8000,
        maxRetries: 1,
        baseDelayMs: 200,
        maxDelayMs: 1000,
        maxConcurrent: 10,
        circuitFailureThreshold: 3,
        circuitHalfOpenAfterMs: 10000,
        circuitSuccessThreshold: 1,
      }
    );

    return { ok: true, model: GEMINI_MODEL, detail: 'Gemini reachable' };
  } catch (error: any) {
    return {
      ok: false,
      model: GEMINI_MODEL,
      detail: `Gemini unreachable: ${error?.message || 'unknown error'}`,
    };
  }
}
